test/test_source_tree_rewriter.rb

```diff
@@ -170,4 +181,70 @@ def test_out_of_range_ranges
     rewriter = Parser::Source::TreeRewriter.new(@buf)
     assert_raises(IndexError) { rewriter.insert_before(range(0, 100), 'hola') }
   end
+
+  def test_empty
+    rewriter = Parser::Source::TreeRewriter.new(@buf)
+    assert_equal true, rewriter.empty?
+
+    # This is a trivial wrap
+    rewriter.wrap(range(2,3), '', '')
+    assert_equal true, rewriter.empty?
+
+    # This is a trivial deletion
+    rewriter.remove(range(2,0))
+    assert_equal true, rewriter.empty?
+
+    rewriter.remove(range(2,3))
+    assert_equal false, rewriter.empty?
+  end
+
+  # splits array into two groups, yield all such possible pairs of groups
+  # each_split([1, 2, 3, 4]) yields [1, 2], [3, 4];
+  #                            then [1, 3], [2, 4]
+  #                                 ...
+  #                     and finally [3, 4], [1, 2]
+  def each_split(array)
+    n = array.size
+    first_split_size = n.div(2)
+    splitting = (0...n).to_set
+    splitting.to_a.combination(first_split_size) do |indices|
+      yield array.values_at(*indices),
+            array.values_at(*(splitting - indices))
+    end
+  end
+
+  # Checks that `actions+extra` give the same result when
+  # made in order or from subgroups that are later merged.
+  # The `extra` actions are always added at the end of the second group.
+  #
+  def check_all_merge_possibilities(actions, extra, **policy)
+    expected = apply(actions + extra, **policy)
+
+    each_split(actions) do |actions_1, actions_2|
```

Comment:
this is error-prone because minitest doesn't complain if there are no assertions in the test. Do you really need to test all combinations? Or it enough to have just a single static test?

Plus, debugging it is a bit complicated if it fails. How many assertions there are? `(N/2)!`?

Reply:
There are `(1..8).to_a.combination(4).size # => 70` possibilities. They are probably not all needed, but it seemed less error prone to test them all than to try to think of what could orders could yield errors...
I've added an error message to help debugging.